[calculations] Detect holes in a V2ObservationResponse and attempt to fill them using calculations

n-h-diaz opened the pull request.

shifucun commented on internal/server/v2/observation/calculation.go:
```diff
@@ -17,56 +17,137 @@ package observation
-					delete(bv, formula)
-					calculatedResponses = append(calculatedResponses, derivedSeries)
-					continue
+				if len(emptyDcids) < len(entity.GetDcids()) {
```
is it possible that len(emptyDcids) > entity.GetDcids()? i wonder if you can remove the check of line 141

n-h-diaz replied:
It's not possible to be greater but it can be equal (i.e. nothing got computed from this round), so I didn't want to keep appending empty responses to result in this case

shifucun commented on internal/server/v2/observation/calculation.go:
```diff
@@ -17,56 +17,137 @@ package observation
+				emptyDcids := []string{}
+				for _, dcid := range entity.GetDcids() {
+					if _, ok := variableObs.ByEntity[dcid]; !ok {
+						emptyDcids = append(emptyDcids, dcid)
```
nit: remove `emptyDcids`

n-h-diaz replied:
Changed to "newEntityDcids"

shifucun commented on internal/server/v2/observation/calculation.go:
```diff
@@ -17,56 +17,137 @@ package observation
+// Detects holes in a V2ObservationResponse and attempts to fill them using calculations.
+func MaybeCalculateHoles(
+	ctx context.Context,
+	store *store.Store,
+	cachedata *cache.Cache,
+	metadata *resource.Metadata,
+	httpClient *http.Client,
```
Now as every function passes in these long list of args, should consolidate them into one class .. for later.

shifucun commented on internal/server/v2/observation/calculation.go:
```diff
@@ -17,56 +17,137 @@ package observation
+	metadata *resource.Metadata,
+	httpClient *http.Client,
+	formula string,
+	variable string,
```
why need `variable`?

n-h-diaz replied:
It's just for constructing the final response (since the formula doesn't reference the target variable anywhere)

shifucun replied:
maybe have a class to wrap the formula and the target variable together?

shifucun commented on internal/server/v2/observation/calculation.go:
```diff
@@ -17,56 +17,137 @@ package observation
+		Filter:   inputReq.GetFilter(),
+		Select:   inputReq.GetSelect(),
+	}
+	variableResp, err := ObservationInternal(ctx, store, cachedata, metadata, httpClient, newReq)
```
s/variableResp/initialObs

n-h-diaz replied:
Done

shifucun commented on internal/server/v2/observation/calculation.go:
```diff
@@ -17,56 +17,137 @@ package observation
+	if err != nil {
+		return nil, err
+	}
+	computeLeafObs(variableResp, variableFormula)
```
now the VariableFormula is not a formula anymore, but the actual tree with observation.

I wonder if it's possible to move `CandidateObs` out of `ASTNode`?

n-h-diaz replied:
Done

shifucun commented on internal/server/v2/observation/calculation.go:
```diff
@@ -17,56 +17,137 @@ package observation
+		return nil, err
+	}
+	computeLeafObs(variableResp, variableFormula)
+	newResp := &pbv2.ObservationResponse{
```
It's not clear why there are both `computeLeafObs()` and `evalExpr(variableFormula.Expr, variableFormula)`

Also this is followed by using calculatedResp to construct final response.

I feel these can be combined and simplfied?

n-h-diaz replied:
evalExpr is a separate function because it's recursive. I also created a separate function for computeLeafObs to only find matching obs once (and then can use these obs throughout the recursion) rather that recomputing each time a node potentially shows up. I also structured it as entity -> facet -> obs to make it easier to match facets (vs ObservationResponse, where facetId is part of the obs), but this means I have to reconstruct the response later. I moved the candidate obs out of the formula (based on your above comment) - let me know if this makes sense, or if you'd prefer I structure it differently!

n-h-diaz commented on internal/server/handler_v2.go:
```diff
@@ -210,11 +210,28 @@ func (s *Server) V2Event(
 	ctx context.Context, in *pbv2.ObservationRequest,
 ) (*pbv2.ObservationResponse, error) {
-	mergedResp, err := v2observation.ObservationInternal(ctx, s.store, s.cachedata.Load(), s.metadata, s.httpClient, in)
+	mergedResp, err := v2observation.ObservationInternal(
```
Done

n-h-diaz commented:
Made some updates based on discussion today, primarily:
* Changed some names to be clearer
* Updated to directly merge ObservationResponses (rather than some other structure), so no need to reconstruct the response
* (Side effect of ^) Preserve order of facets

Please take a look when you get the chance 😃

shifucun commented on internal/server/v2/observation/calculation_util.go:
```diff
@@ -212,90 +211,143 @@ func compareFacet(facet1, facet2 *pb.Facet) bool {
+	if len(x.ByVariable) != 1 || len(y.ByVariable) != 1 {
+		return nil, fmt.Errorf("more than one variable in intermediate response")
+	}
+	for _, xVariableObs := range x.ByVariable {
+		for _, yVariableObs := range y.ByVariable {
```
no need to have a for loop x.ByVariable and y.ByVariable if the length is 1 (from line 304)

n-h-diaz replied:
Done

shifucun commented on internal/server/v2/observation/calculation_util.go:
```diff
@@ -212,90 +211,143 @@ func compareFacet(facet1, facet2 *pb.Facet) bool {
+					continue
+				}
+				newOrderedFacets := []*pbv2.FacetObservation{}
+				for i := 0; i < len(xEntityObs.OrderedFacets); i++ {
```
after line 314, can define two short-name alias and use them below

xFacets := xEntityObs.OrderedFacets
yFacets := yEntityObs.OrderedFacets

n-h-diaz replied:
Done

shifucun commented on internal/server/v2/observation/calculation_util.go:
```diff
@@ -212,90 +211,143 @@ func compareFacet(facet1, facet2 *pb.Facet) bool {
+				for i := 0; i < len(xEntityObs.OrderedFacets); i++ {
+					for j := 0; j < len(yEntityObs.OrderedFacets); j++ {
+						if xEntityObs.OrderedFacets[i].GetFacetId() == yEntityObs.OrderedFacets[j].GetFacetId() {
+							newFacetId := xEntityObs.OrderedFacets[i].GetFacetId()
```
so merged facet id is the same as x? Need a todo to figure out full solution for new facet id?

n-h-diaz replied:
Right now I'm just using the same facet and keeping all the fields the same. I think we can maybe also update to use dcAggregate/ + maybe change the import name (to match what's done now in pipeline), but I need to figure out how to update the facetId accordingly -- added a TODO for this
